File: src/main/java/com/ocbc/auctionservice/loggings/LoggingDiagnosticFilter.java
package com.ocbc.auctionservice.loggings;

import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;
import org.springframework.core.annotation.Order;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.UUID;

import static java.lang.String.format;

@Slf4j
@Order(1)
public class LoggingDiagnosticFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain
    ) throws ServletException, IOException {
        MDC.clear();

        Long startTime = System.currentTimeMillis();
        try {
            MDC.putCloseable("requestId", getRequestId(request));
            log.info(format("Starting request %s", request.getRequestURI()));
            filterChain.doFilter(request, response);
        } finally {
            Long endTime = System.currentTimeMillis();
            Long duration = endTime - startTime;
            MDC.putCloseable("responseCode", response.getStatus() + "");
            MDC.putCloseable("responseTimeMillis", duration + "");
            log.info(format("Completed Request %s", request.getRequestURI()));
            log.info(format("Completed Request in %s duration", duration + ""));
        }
    }

    private String getRequestId(HttpServletRequest request) {
        String header = request.getHeader("X-B3-TraceId");
        if (header != null) {
            return header;
        }
        return UUID.randomUUID().toString();
    }
}

File: src/main/kotlin/com/auction/service/logging/LoggingDiagnosticFilter.kt
package com.auction.service.logging

import mu.KotlinLogging
import org.slf4j.MDC
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class LoggingDiagnosticFilter : OncePerRequestFilter() {

    private val logger = KotlinLogging.logger {}

    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain
    ) {
        MDC.clear()
        val startTime = System.currentTimeMillis()
        try {
            MDC.putCloseable("requestId", getRequestId(request))
            logger.info("[START] request ${request.requestURI}")
            filterChain.doFilter(request, response)
        } finally {
            val endTime = System.currentTimeMillis()
            val duration = endTime - startTime
            MDC.putCloseable("responseCode", response.status.toString() + "")
            MDC.putCloseable("responseTimeMillis", duration.toString() + "")
            logger.info("[COMPLETE] Request ${request.requestURI} in $duration duration")
        }
    }

    private fun getRequestId(request: HttpServletRequest): String {
        val header = request.getHeader("X-B3-TraceId")
        return header ?: UUID.randomUUID().toString()
    }
}